Tests/SourceKitTests/BuildSystemTests.swift

```diff
@@ -0,0 +1,167 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SourceKit
+import Basic
+import LanguageServerProtocol
+import SKCore
+import SKSupport
+import SKTestSupport
+import XCTest
+
+// Workaround ambiguity with Foundation.
+typealias LSPNotification = LanguageServerProtocol.Notification
+
+/// Build system to be used for testing BuildSystem and BuildSystemDelegate functionality with SourceKitServer
+/// and other components.
+final class TestBuildSystem: BuildSystem {
+  var indexStorePath: AbsolutePath? = nil
+  var indexDatabasePath: AbsolutePath? = nil
+
+  weak var delegate: BuildSystemDelegate?
+
+  /// Build settings by file.
+  var buildSettingsByFile: [URL: FileBuildSettings] = [:]
+
+  /// Toolchains by file.
+  var toolchainsByFile: [URL: Toolchain] = [:]
+
+  /// Files currently being watched by our delegate.
+  var watchedFiles: Set<URL> = []
+
+  func settings(for url: URL, _ language: Language) -> FileBuildSettings? {
+    return buildSettingsByFile[url]
+  }
+
+  func toolchain(for url: URL, _ language: Language) -> Toolchain? {
+    return toolchainsByFile[url]
+  }
+
+  func registerChangeWatching(for url: URL) {
+    watchedFiles.insert(url)
+  }
+
+  func unregisterChangeWatching(for url: URL) {
+    watchedFiles.remove(url)
+  }
+}
+
+final class BuildSystemTests: XCTestCase {
+
+  /// Connection and lifetime management for the service.
+  var testServer: TestSourceKitServer! = nil
+
+  /// The primary interface to make requests to the SourceKitServer.
+  var sk: TestClient! = nil
+
+  /// The server's workspace data. Accessing this is unsafe if the server does so concurrently.
+  var workspace: Workspace! = nil
+
+  /// The build system that we use to verify SourceKitServer behavior.
+  var buildSystem: TestBuildSystem! = nil
+
+  override func setUp() {
+    testServer = TestSourceKitServer()
+    buildSystem = TestBuildSystem()
+
+    self.workspace = Workspace(
+      rootPath: nil,
+      clientCapabilities: ClientCapabilities(),
+      buildSettings: buildSystem,
+      index: nil,
+      buildSetup: TestSourceKitServer.buildSetup)
+    testServer.server!.workspace = workspace
+
+    sk = testServer.client
+    _ = try! sk.sendSync(InitializeRequest(
+        processId: nil,
+        rootPath: nil,
+        rootURL: nil,
+        initializationOptions: nil,
+        capabilities: ClientCapabilities(workspace: nil, textDocument: nil),
+        trace: .off,
+        workspaceFolders: nil))
+  }
+
+  override func tearDown() {
+    buildSystem = nil
+    workspace = nil
+    sk = nil
+    testServer = nil
+  }
+
+  func testDocumentUpdatedBuildSettings() {
+    let url = URL(fileURLWithPath: "/a.swift")
+
+    sk.sendNoteSync(DidOpenTextDocument(textDocument: TextDocumentItem(
+      url: url,
+      language: .swift,
+      version: 12,
+      text: """
+      func
```

Comment:
To test that the settings update is working correctly, I suggest using `#if FOO` and changing the compiler arguments to add/remove `-DFOO`.

Reply:
Hmm this doesn't seem to be working, do I need to add in other Swift toolchain flags?

Reply:
I would expect it to work just like a command-line invocation of swiftc.  E.g.

```
$ cat t.swift
#if FOO
func foo() {}
#endif

foo()

$ swiftc t.swift
t.swift:5:1: error: use of unresolved identifier 'foo'
foo()
^~~

$ swiftc t.swift -DFOO
# OK
```

Reply:
With just the build settings of `-DFOO` I'm not seeing any diagnostics reported by SourceKit even with the above code

Reply:
In that example the diagnostics should be there when it is *not* defined.

Reply:
I'm not seeing any diagnostics in the Swift code although it's working fine for ObjC

Reply:
Ah, sorry: you need to expect two notifications in Swift, because the syntactic errors (if any) come back immediately, but the semantic ones show up when the AST finishes typechecking/etc.  You can see the way I've handled this in `LocalSwiftTests.testEditing`, but basically you just pass a second handler.  I think you want to just have the first handler be empty, then do the check in the second one where you know you have semantic errors.

Reply:
Interesting, I'm not seeing a second notification, presumably because SourceKit doesn't have the proper arguments for semantic analysis? I'll try mirroring the flags from the fallback build system.

Reply:
Works with the FallbackBuildSystem